src/cryptokit.ml

```diff
@@ -2072,6 +2074,98 @@ let new_key ?(rng = Random.secure_rng) ?e numbits =
+      n: string;
+      n2: string;
+      g: string;
+      p: string;
+      q: string;
+      lambda: string;
+      mu: string}
+  let wipe_key k =
+    wipe_string k.n;
+    wipe_string k.n2;
+    wipe_string k.g;
+    wipe_string k.p;
+    wipe_string k.q;
+    wipe_string k.lambda;
+    wipe_string k.mu
+
+  let encrypt key msg ?r ()=
+    let check_random_number r =
+      if (Bn.(relative_prime r (Bn.of_bytes key.n)) && r<(Bn.of_bytes key.n)) then true else false in
+    let rec get_r () =
+      let r = Bn.random ~rng:(Random.secure_rng#random_bytes) (key.size-1) in
+      if check_random_number r then r else get_r () in
+    let r = match r with
+      | None -> get_r ()
+      | Some r -> if check_random_number r then r else raise(Error Wrong_random_number )in
+    let msg = Bn.of_bytes msg in
+    let n = Bn.of_bytes key.n in
+    let n2 = Bn.of_bytes key.n2 in
+    let g = Bn.of_bytes key.g in
+    if Bn.(compare msg zero) <= 0 then raise (Error Negative_msg);
+    let gm = Bn.mod_power g msg n2 in
+    let rn = Bn.mod_power r n n2 in
+    let c = Bn.(mod_ (mult gm rn) n2) in
+    let c_bytes = Bn.to_bytes c in
+    Bn.wipe msg; Bn.wipe n; Bn.wipe n2; Bn.wipe g;
+    Bn.wipe r; Bn.wipe gm; Bn.wipe rn; Bn.wipe c;
+    c_bytes
+
+  let decrypt key c =
+    let c = Bn.of_bytes c in
+    let n = Bn.of_bytes key.n in
+    let n2 = Bn.of_bytes key.n2 in
+    let lambda = Bn.of_bytes key.lambda in
+    let mu = Bn.of_bytes key.mu in
+    let cn = Bn.mod_power c lambda n2 in
+    let lx = Bn.(div (sub cn one) n) in
+    let m = Bn.(mod_ (mult lx mu) n) in
+    let msg = Bn.to_bytes m in
+    Bn.wipe c; Bn.wipe n; Bn.wipe n2; Bn.wipe lambda;
+    Bn.wipe mu; Bn.wipe cn; Bn.wipe lx; Bn.wipe m;
+    msg
+
+  let new_key ?(rng = Random.secure_rng) numbits =
+    if numbits < 32 || numbits land 1 > 0 then raise(Error Wrong_key_size);
+    let numbits2 = numbits / 2 in
+    (* Make sure p > q *)
+    let rec gen_factors nbits =
+      let p = Bn.random_prime ~rng:(rng#random_bytes) nbits
+      and q = Bn.random_prime ~rng:(rng#random_bytes) nbits in
+      let cmp = Bn.compare p q in
+      if cmp = 0 then gen_factors nbits else
+      if Bn.(relative_prime (mult p q) (mult (sub p one) (sub q one))) then (p,q) else
+      (* Recursively generate factors *)
+      gen_factors nbits in
+    let (p, q) = gen_factors numbits2 in
+    (* n = pq *)
+    let n = Bn.mult p q in
+    let n2 = Bn.mult n n in
+    let g = Bn.add n Bn.one in
+    let lambda = Bn.(lcm (sub p one) (sub q one)) in
+    let mu = Bn.mod_inv lambda n in
+
+    (* Build key *)
+    let res =
+      { size = numbits;
+        n = Bn.to_bytes ~numbits:numbits n;
+        n2 = Bn.to_bytes n2;
+        g = Bn.to_bytes g;
+        p = Bn.to_bytes ~numbits:numbits2 p;
+        q = Bn.to_bytes ~numbits:numbits2 q;
+        lambda = Bn.to_bytes lambda;
+        mu = Bn.to_bytes mu} in
+    Bn.wipe n; Bn.wipe n2; Bn.wipe g;
+    Bn.wipe p; Bn.wipe q;
+    Bn.wipe lambda; Bn.wipe mu;
```

Comment:
Note that `sub p one` and `sub q one` should probably be computed only once so that they can be wiped at the end.

Reply:
Done :)